File: aoc2025/javascript/grid.js
import { readFile } from './utilities.js';

// Constants
export const EMPTY = '.';
export const PAPER_ROLL = '@';
export const MOVED_UP = '^';
export const MOVED_DOWN = 'v';
export const MOVED_LEFT = '<';
export const MOVED_RIGHT = '>';
export const START = 'S';
export const TACHYON = MOVED_UP;

// Direction constants
export const RIGHT = { x: 1, y: 0 };
export const LEFT = { x: -1, y: 0 };
export const DOWN = { x: 0, y: 1 };
export const UP = { x: 0, y: -1 };
export const RIGHT_DOWN = { x: 1, y: 1 };
export const RIGHT_UP = { x: 1, y: -1 };
export const LEFT_DOWN = { x: -1, y: 1 };
export const LEFT_UP = { x: -1, y: -1 };

export const DIRECTIONS = [LEFT, RIGHT, UP, DOWN, LEFT_DOWN, LEFT_UP, RIGHT_DOWN, RIGHT_UP];

export class Point {
  constructor(x, y) {
    this.x = x;
    this.y = y;
  }

  equals(other) {
    return this.x === other.x && this.y === other.y;
  }

  toString() {
    return `Point(${this.x}, ${this.y})`;
  }
}

export class Visit {
  constructor(point, direction) {
    this.point = point;
    this.direction = direction;
  }
}

export class Grid {
  constructor(filename) {
    const lines = readFile(filename);
    this.height = lines.length;
    this.width = lines.length > 0 ? lines[0].length : 0;
    this.map = [];
    this.visited = [];
    this.start = null;
    this.current = null;
    this.tiles = {};

    for (let i = 0; i < lines.length; i++) {
      this.map[i] = [];
      for (let j = 0; j < lines[i].length; j++) {
        const char = lines[i][j];
        this.map[i][j] = char;
        if (char === START) {
          this.start = new Point(j, i);
          this.current = new Point(j, i);
          this.visited.push(new Visit(new Point(j, i), { x: 0, y: 0 }));
        }
      }
    }
  }

  static makeGridByPoints(sizePoints, ...tiles) {
    let maxX = 0;
    let maxY = 0;
    for (const point of sizePoints) {
      if (point.x > maxX) maxX = point.x;
      if (point.y > maxY) maxY = point.y;
    }

    const grid = new Grid.__emptyGrid();
    grid.width = maxX + 1;
    grid.height = maxY + 1;
    grid.map = Array.from({ length: maxY + 1 }, () => Array(maxX + 1).fill(EMPTY));
    grid.tiles = {};
    grid.visited = [];

    for (const tileDef of tiles) {
      grid.tiles[tileDef.char] = tileDef.color;
      for (const point of tileDef.points) {
        grid.map[point.y][point.x] = tileDef.char;
      }
    }

    return grid;
  }

  static __emptyGrid() {
    const grid = Object.create(Grid.prototype);
    return grid;
  }

  setStart(point) {
    this.start = point;
    this.current = point;
    this.visited = [new Visit(point, { x: 0, y: 0 })];
  }

  setObject(point, object) {
    this.map[point.y][point.x] = object;
  }

  positionContainsObject(point, object) {
    if (!this.pointWithinBounds(point)) {
      return false;
    }
    return this.map[point.y][point.x] === object;
  }

  pointWithinBounds(point) {
    return !(point.x < 0 || point.x >= this.width || point.y < 0 || point.y >= this.height);
  }

  canMoveRight() {
    return this.current.x < this.width - 1;
  }

  canMoveLeft() {
    return this.current.x > 0;
  }

  canMoveDown() {
    return this.current.y < this.height - 1;
  }

  canMoveUp() {
    return this.current.y > 0;
  }

  canMove(direction) {
    if (direction === RIGHT || (direction.x === 1 && direction.y === 0)) {
      return this.canMoveRight();
    }
    if (direction === LEFT || (direction.x === -1 && direction.y === 0)) {
      return this.canMoveLeft();
    }
    if (direction === DOWN || (direction.x === 0 && direction.y === 1)) {
      return this.canMoveDown();
    }
    if (direction === UP || (direction.x === 0 && direction.y === -1)) {
      return this.canMoveUp();
    }

    let canMoveX = true;
    let canMoveY = true;

    if (direction.x > 0) {
      canMoveX = this.canMoveRight();
    } else if (direction.x < 0) {
      canMoveX = this.canMoveLeft();
    }

    if (direction.y > 0) {
      canMoveY = this.canMoveDown();
    } else if (direction.y < 0) {
      canMoveY = this.canMoveUp();
    }

    return canMoveX && canMoveY;
  }

  move(direction) {
    const canMove = this.canMove(direction);

    if (!canMove) {
      return false;
    }

    const point = new Point(
      this.current.x + direction.x,
      this.current.y + direction.y
    );

    const visit = new Visit(point, direction);
    this.visited.push(visit);
    this.current = point;

    return true;
  }

  findLastObjectToRight(start, object) {
    for (let x = this.width - 1; x > start.x; x--) {
      if (this.map[start.y][x] === object) {
        return new Point(x, start.y);
      }
    }
    return new Point(0, 0);
  }

  findLastObjectToBottom(start, object) {
    for (let y = this.height - 1; y > start.y; y--) {
      if (this.map[y][start.x] === object) {
        return new Point(start.x, y);
      }
    }
    return new Point(0, 0);
  }

  print() {
    for (const row of this.map) {
      console.log(row.join(''));
    }
  }

  printVisited() {
    const gridCopy = this.map.map(row => [...row]);
    const visitMap = new Map();

    for (const visit of this.visited) {
      gridCopy[visit.point.y][visit.point.x] = this._directionToArrow(visit.direction);
      visitMap.set(`${visit.point.x},${visit.point.y}`, true);
    }

    for (let y = 0; y < gridCopy.length; y++) {
      console.log(gridCopy[y].join(''));
    }
  }

  _directionToArrow(direction) {
    if (direction.x === 0 && direction.y === 0) {
      return START;
    }
    if (direction.x === 0 && direction.y === -1) {
      return MOVED_UP;
    }
    if (direction.x === 0 && direction.y === 1) {
      return MOVED_DOWN;
    }
    if (direction.x === -1 && direction.y === 0) {
      return MOVED_LEFT;
    }
    if (direction.x === 1 && direction.y === 0) {
      return MOVED_RIGHT;
    }
    return 'X';
  }
}

File: aoc2025/zig/src/grid.zig
const std = @import("std");
const file_parser = @import("file_parser.zig");

pub const EMPTY = '.';
pub const PAPER_ROLL = '@';
pub const MOVED_UP = '^';
pub const MOVED_DOWN = 'v';
pub const MOVED_LEFT = '<';
pub const MOVED_RIGHT = '>';
pub const START = 'S';
pub const TACHYON = MOVED_UP;

pub const Point = struct {
    x: i32,
    y: i32,
    
    pub fn init(x: i32, y: i32) Point {
        return Point{ .x = x, .y = y };
    }
};

pub const RIGHT = Point{ .x = 1, .y = 0 };
pub const LEFT = Point{ .x = -1, .y = 0 };
pub const DOWN = Point{ .x = 0, .y = 1 };
pub const UP = Point{ .x = 0, .y = -1 };
pub const RIGHT_DOWN = Point{ .x = 1, .y = 1 };
pub const RIGHT_UP = Point{ .x = 1, .y = -1 };
pub const LEFT_DOWN = Point{ .x = -1, .y = 1 };
pub const LEFT_UP = Point{ .x = -1, .y = -1 };

pub const DIRECTIONS = [_]Point{
    LEFT, RIGHT, UP, DOWN, LEFT_DOWN, LEFT_UP, RIGHT_DOWN, RIGHT_UP,
};

pub const Visit = struct {
    point: Point,
    direction: Point,
};

pub const Grid = struct {
    width: usize,
    height: usize,
    start: Point,
    current: Point,
    visited: []Visit,
    visited_count: usize,
    visited_capacity: usize,
    map: [][]u8,
    allocator: std.mem.Allocator,
    
    pub fn init(allocator: std.mem.Allocator, filename: []const u8) !Grid {
        const lines = try file_parser.readFile(allocator, filename);
        defer allocator.free(lines);
        
        if (lines.len == 0) {
            return error.EmptyFile;
        }
        
        const width = lines[0].len;
        const height = lines.len;
        
        var map = try allocator.alloc([]u8, height);
        for (0..height) |y| {
            map[y] = try allocator.alloc(u8, width);
            for (0..width) |x| {
                map[y][x] = lines[y][x];
            }
        }
        
        var start = Point{ .x = 0, .y = 0 };
        for (0..height) |y| {
            for (0..width) |x| {
                if (map[y][x] == START) {
                    start = Point{ .x = @intCast(x), .y = @intCast(y) };
                    break;
                }
            }
        }
        
        const visited_capacity = 16;
        var visited = try allocator.alloc(Visit, visited_capacity);
        visited[0] = Visit{ .point = start, .direction = Point{ .x = 0, .y = 0 } };
        
        return Grid{
            .width = width,
            .height = height,
            .start = start,
            .current = start,
            .visited = visited,
            .visited_count = 1,
            .visited_capacity = visited_capacity,
            .map = map,
            .allocator = allocator,
        };
    }
    
    pub fn deinit(self: *Grid) void {
        for (self.map) |row| {
            self.allocator.free(row);
        }
        self.allocator.free(self.map);
        self.allocator.free(self.visited);
    }
    
    pub fn setStart(self: *Grid, point: Point) void {
        self.start = point;
        self.current = point;
        self.visited_count = 0;
        self.visited[0] = Visit{ .point = point, .direction = Point{ .x = 0, .y = 0 } };
        self.visited_count = 1;
    }
    
    pub fn setObject(self: *Grid, point: Point, object: u8) void {
        if (self.pointWithinBounds(point)) {
            self.map[@as(usize, @intCast(point.y))][@as(usize, @intCast(point.x))] = object;
        }
    }
    
    pub fn positionContainsObject(self: *Grid, point: Point, object: u8) bool {
        if (!self.pointWithinBounds(point)) {
            return false;
        }
        return self.map[@as(usize, @intCast(point.y))][@as(usize, @intCast(point.x))] == object;
    }
    
    pub fn pointWithinBounds(self: *Grid, point: Point) bool {
        return point.x >= 0 and point.x < @as(i32, @intCast(self.width)) and
               point.y >= 0 and point.y < @as(i32, @intCast(self.height));
    }
    
    pub fn canMove(self: *Grid, direction: Point) bool {
        const newPoint = Point{
            .x = self.current.x + direction.x,
            .y = self.current.y + direction.y,
        };
        return self.pointWithinBounds(newPoint);
    }
    
    pub fn move(self: *Grid, direction: Point) !bool {
        if (!self.canMove(direction)) {
            return false;
        }
        
        const newPoint = Point{
            .x = self.current.x + direction.x,
            .y = self.current.y + direction.y,
        };
        
        const visit = Visit{
            .point = newPoint,
            .direction = direction,
        };
        
        if (self.visited_count >= self.visited_capacity) {
            const new_visited = try self.allocator.realloc(self.visited, self.visited_capacity * 2);
            self.visited = new_visited;
            self.visited_capacity *= 2;
        }
        
        self.visited[self.visited_count] = visit;
        self.visited_count += 1;
        self.current = newPoint;
        
        return true;
    }
    
    pub fn print(self: *Grid) void {
        for (0..self.height) |y| {
            for (0..self.width) |x| {
                std.debug.print("{c}", .{self.map[y][x]});
            }
            std.debug.print("\n");
        }
    }
};